sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/DeploymentStep.java

```diff
@@ -176,25 +177,48 @@ private void setTaskStatus(Protos.TaskID taskID, Status status) {
         }
     }
 
-    private Status getStatus(Map<Protos.TaskID, TaskStatusPair> tasks) {
+    @VisibleForTesting
+    Status getStatus(Map<Protos.TaskID, TaskStatusPair> tasks) {
         if (tasks.isEmpty()) {
             return Status.PENDING;
         }
 
+        Set<Status> statuses = new HashSet<>();
         for (Map.Entry<Protos.TaskID, TaskStatusPair> entry : tasks.entrySet()) {
             String taskId = entry.getKey().getValue();
             Status status = entry.getValue().getStatus();
             logger.info("TaskId: {} has status: {}", taskId, status);
-            if (!status.equals(Status.COMPLETE)) {
-                // Keep and log current status
-                return super.getStatus();
-            }
+
+            statuses.add(status);
+        }
+
+        // A DeploymentStep should have the "least" status of its consituent tasks.
+        // We can rely on the order of Status as returned by its getOrder method.
+        // For example:
+        // 1 PENDING, 2 STARTING => PENDING
+        // 2 STARTING, 1 COMPLETE => STARTING
+        // 3 COMPLETE => COMPLETE
+        if (statuses.contains(Status.ERROR)) {
+            return Status.ERROR;
+        } else if (statuses.contains(Status.PENDING)) {
+            return Status.PENDING;
+        } else if (statuses.contains(Status.PREPARED)) {
+            return Status.PREPARED;
+        } else if (statuses.contains(Status.STARTING)) {
+            return Status.STARTING;
+        } else if (statuses.contains(Status.COMPLETE) && statuses.size() == 1) {
+            return Status.COMPLETE;
```

Comment:
OK. Was a bit confused, but then I realised this is equivalent to **ALL** tasks being `COMPLETE`. We could use a `Map<Status, Integer>` instead of a set and count the number of tasks in each state, and then only return `COMPLETE` if the number of tasks in this state is the total number of tasks.

Reply:
I could add a comment if you like. I prefer set over map. Less complex.